### PR TITLE
Guard cache lifecycle methods by mutex

### PR DESCRIPTION
**Pull request by vbekiaris**
`Cache` create & destroy methods are now guarded by mutexes so that creation & destruction are now executed atomically per `cacheNameWithPrefix`.

Requires an EE counterpart - https://github.com/hazelcast/hazelcast-enterprise/pull/1633

(edited to reflect changed scope of pull request)

### CODE REVIEW

**Comment by mmedenjak** [comment]
run-lab-run

**Comment by vbekiaris** [comment]
run-lab-run

**Comment by vbekiaris** [comment]
verify

**Comment by vbekiaris** [comment]
@Donnerbart , @mmedenjak thanks for the reviews. The scope of this PR was changed (see new description) and now also includes new concurrent create/destroy tests, could you have another look?

**Comment by vbekiaris** [comment]
Fixes #11259

File: hazelcast/src/main/java/com/hazelcast/cache/impl/HazelcastServerCacheManager.java [hunk 1]
```diff
@@ -177,6 +180,33 @@ private void enableStatisticManagementOnOtherNodes(String cacheName, boolean sta
     }
 
     @Override
+    public <K, V, C extends Configuration<K, V>> ICache<K, V> createCache(String cacheName, C configuration)
```
[the remark this answers is not in the document]
Comment (vbekiaris) [reply hazelcast/src/main/java/com/hazelcast/cache/impl/HazelcastServerCacheManager.java]: In the context of this pull request I am only dealing with create/destroy races on the member side which may result in partially constructed or disposed-of data structures.

File: hazelcast/src/main/java/com/hazelcast/cache/impl/HazelcastServerCacheManager.java [hunk 2]
```diff
@@ -187,12 +217,6 @@ private void enableStatisticManagementOnOtherNodes(String cacheName, boolean sta
     }
 
     @Override
-    protected void removeCacheConfigFromLocal(String cacheName) {
-        cacheService.deleteCacheConfig(cacheName);
```
[the remark this answers is not in the document]
Comment (vbekiaris) [reply hazelcast/src/main/java/com/hazelcast/cache/impl/HazelcastServerCacheManager.java]: Actually the motivation to remove this overridden method was that in the `cache.destroy` path it is redundant (the `CacheConfig` has already been removed in `AbstractCacheService.destroyDistributedObject`). However I didn't consider direct invocation of `HazelcastCacheManager.removeCache`, so I will restore it, well spotted.

File: hazelcast/src/main/java/com/hazelcast/cache/impl/AbstractCacheService.java [hunk 3]
```diff
@@ -98,6 +98,8 @@ public CacheStatisticsImpl createNew(String name) {
             };
     // mutex factory ensures each Set<Closeable> of cache resources is only constructed and inserted in resources map once
     protected final ContextMutexFactory cacheResourcesMutexFactory = new ContextMutexFactory();
+    // protect concurrent destruction of same Cache
+    protected final ContextMutexFactory cacheLifecycleMutexFactory = new ContextMutexFactory();
```
[the remark this answers is not in the document]
Comment (vbekiaris) [reply hazelcast/src/main/java/com/hazelcast/cache/impl/AbstractCacheService.java]: Right, will update the comment accordingly

**Comment by vbekiaris** [comment]
@mmedenjak updated

**Pull request merged**